clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -196,6 +197,44 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
+bool isArraySizedZero(QualType Ty) {
+  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty)) {
+    const llvm::APInt size = CATy->getSize();
+    return size == 0;
+  }
+  return false;
+}
+
+void Sema::CheckVarDeclOKIfInKernel(VarDecl *var) {
+  // not all variable types supported in kernel contexts
+  // if not we record a deferred diagnostic.
+  if (getLangOpts().SYCLIsDevice) {
+    QualType Ty = var->getType();
+    SourceRange Loc = var->getLocation();
+
+    // __int128, __int128_t, __uint128_t
+    if (Ty->isSpecificBuiltinType(BuiltinType::Int128) ||
+        Ty->isSpecificBuiltinType(BuiltinType::UInt128))
+      SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_type_unsupported)
+          << Ty.getUnqualifiedType().getCanonicalType().getAsString();
+
+    // QuadType __float128
+    if (Ty->isSpecificBuiltinType(BuiltinType::Float128) &&
+        !Context.getTargetInfo().hasFloat128Type())
+      SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_type_unsupported)
+          << "__float128";
+
+    // zero length arrays
+    if (Ty->isArrayType() && isArraySizedZero(Ty))
```

Comment:
the RHS of the && already checks the first predicate, no need to check both here.

Reply:
done